node/src/components/consensus/protocols/highway.rs

```diff
@@ -283,6 +283,20 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
                         sender,
                     ));
                 } else {
+                    // If unit is sent by a doppelganger, deactivate this instance of an active
+                    // validator. Continue processing the unit so that it can be added to the state.
+                    if let Some(swunit) = pvv.inner().signed_wire_unit() {
+                        if self.highway.is_doppelganger(swunit) {
```

Comment:
Is it guaranteed at this point that the unit isn't already in the state, and someone just re-sent our own unit back to us? (Maybe it should be part of `is_doppelganger` after all; it's a cheap check.)

Reply:
The unit isn't in the state but I guess you're right that someone could have send it back to us. Especially that `state.has_vertex(v)` check is made much later. Maybe we should rearrange that and verify whether the vertex is already stored much earlier? Even before `pre_validate` and `validate` checks? Testing membership should be much faster than those two.